File: draft_design_pattern/composite.py
# どんな組織も従業員（employee）で構成されているものです。
# 各従業員は同じ機能（給与や責務など）を共有しており、
# 従業員によっては報告の義務があったりなかったり、部下がいたりいなかったりします。
# compositeパターンは、クライアントが個別のオブジェクトを統一的な方法で扱えるようにするためのものです。
from abc import ABC, abstractmethod


class Employee(ABC):
    def __init__(self, name: str, salary: float) -> None:
        self.name = name
        self.salary = salary

    @abstractmethod
    def getName(self) -> str:
        pass

    @abstractmethod
    def setSalary(self, salary: float) -> None:
        pass

    @abstractmethod
    def getSalary(self) -> float:
        pass

    @abstractmethod
    def getRoles(self) -> list["Employee"]:
        pass


class Developer(Employee):
    def __init__(self, name: str, salary: float) -> None:
        super().__init__(name, salary)
        self.roles: list["Employee"] = []

    def getName(self) -> str:
        return self.name

    def setSalary(self, salary: float) -> None:
        self.salary = salary

    def getSalary(self) -> float:
        return self.salary

    def getRoles(self) -> list["Employee"]:
        return self.roles


class Designer(Employee):
    def __init__(self, name: str, salary: float) -> None:
        super().__init__(name, salary)
        self.roles: list["Employee"] = []

    def getName(self) -> str:
        return self.name

    def setSalary(self, salary: float) -> None:
        self.salary = salary

    def getSalary(self) -> float:
        return self.salary

    def getRoles(self) -> list["Employee"]:
        return self.roles


# 次に、さまざまな種別の従業員で構成された組織Organizationを記述します


class Organization:
    def __init__(self) -> None:
        self.employees: list[Employee] = []

    def addEmployee(self, employee: Employee):
        self.employees.append(employee)

    def getNetSalaries(self) -> float:
        netSalary: float = 0
        for v in self.employees:
            netSalary += v.getSalary()
        return netSalary


# 従業員を準備
john = Developer('John Doe', 12000)
jane = Designer('Jane Doe', 15000)

# 組織に追加
organization = Organization()
organization.addEmployee(john)
organization.addEmployee(jane)

print(f"正味の給与: {organization.getNetSalaries()}")

File: draft_design_pattern/composite.nim
type
  Employee = ref object of RootObj
    name: string
    salary: float
    roles: seq[Employee]

  Developer = ref object of Employee
  Designer = ref object of Employee

  Organization = ref object
    employees: seq[Employee]


method getName(self: Employee): string {.base, locks: "unknown".} =
  discard

method setSalary(self: Employee, salary: float) {.base, locks: "unknown".} =
  discard

method getSalary(self: Employee): float {.base, locks: "unknown".} =
  discard

method getRoles(self: Employee): seq[Employee] {.base, locks: "unknown".} =
  discard


method getName(self: Developer): string {.locks: "unknown".} =
  self.name

method setSalary(self: Developer, salary: float) {.locks: "unknown".} =
  self.salary = salary

method getSalary(self: Developer): float {.locks: "unknown".} =
  self.salary

method getRoles(self: Developer): seq[Employee] {.locks: "unknown".} =
  self.roles


method getName(self: Designer): string {.locks: "unknown".} =
  self.name

method setSalary(self: Designer, salary: float) {.locks: "unknown".} =
  self.salary = salary

method getSalary(self: Designer): float {.locks: "unknown".} =
  self.salary

method getRoles(self: Designer): seq[Employee] {.locks: "unknown".} =
  self.roles


proc addEmployee(self: Organization, employee: Employee) =
  self.employees.add(employee)

proc getNetSalaries(self: Organization): float =
  result = 0
  for v in self.employees:
    result += v.getSalary()


proc main() =
  let john = Developer(name: "John Doe", salary: 12000)
  let jane = Designer(name: "Jane Doe", salary: 15000)

  let organization = Organization()
  organization.addEmployee(john)
  organization.addEmployee(jane)

  echo "正味の給与: " & $organization.getNetSalaries()


main()
